design/m1_push_to_file_design.md

```diff
@@ -0,0 +1,1280 @@
+| 8 | Unable to retrieve Conjur variables | You are using SP in "file" mode | | Error: Failed to provide DAP/Conjur secrets<br><br>SP fails | CSPFK016E | Unit | |
+| 9 | Missing volume mounts | You are using SP in "file" mode | One or more volumes cannot be found by Secrets Provider. | Error: Unable to access volume "{value}". Error: "{value}". Ensure that the correct volumes are defined in the deployment manifest and attached to the Secrets Provider container.<br><br>SP fails to start | | Unit | US-07 |
+| 10 | File permissions error | You are using SP in "file" mode | Secrets Provider does not have permission to write a secrets file to the specified volume and mount path. | Error: Secrets Provider does not have permission to write the secrets file "{value}". Please check the volume mount permissions on the Secrets Provider container and try again.<br><br>SP fails | | Unit | US-07 |
+| 11 | Invalid file path provided | You are using SP in "file" mode and have provided a value for the conjur.org/secret-file-path annotation | The provided file path is not a valid file path. | Error: You have provided an invalid file path "{value}" for the "{value}" secret group.<br><br>SP fails to start | | Unit | US-07 |
+| 12 | File conflict between two secret groups | You are using SP in "file" mode and have defined multiple secret groups | One or more secret groups have he same ouput file. | Error: You have provided conflicting file paths; secret groups "{value}", "{value}", ... all have output file "{value}".<br><br>SP fails to start | | Unit | US-07 |
+
+#### Integration / E2E tests
+
+E2E tests are expensive. We should try to limit their numbers. There are cheaper ways to validate behavior. For the push to file functionality it will likely be possible to validate the happy path in one fell swoop by using configuration tht captures all the supported annotations and template types.
+
+- [ ] Happy path, Secret Provider runs as an init container in file mode and successfully injects files. This single test can cover all supported annotations and template types.
+
+The sad path is where things tend to become difficult, because there's a myriad of ways in which failure can occur. A possible way to reduce the need to E2E test failures is to establish a nexus point for all failures in the code so that an error can be validated at the unit/functional testing level.
+
+- [ ] Sad path, Secret Provider runs as an init container in file mode and fails, thereby preventing the deployment from proceeding. TBC it's possible we don't need this at all.
+
+#### Security testing
+
+Security testing will include:
+
+- Automated Vulnerability scans
```

Comment:
This probably depends heavily on how the secrets files are consumed by the application.

For `yaml` or `json` file formats, I don't know if there are best practices that we can recommend e.g. to escape secrets values or the file content to avoid code injection or XSS?

For `bash` or `dotenv`, we can probably recommend that the application does not consume these using `eval`.

For custom templates, customers will be providing their own templates and will be rendering the templates using their own Conjur secrets values. In this case, there isn't much control over what they implement... unless we implemented some highly complex mechanism to look for potential security vulnerability patterns in the templates that are provided. We can at least document some best practices e.g. for escaping HTML text before writing secrets file content to disk.